radixdlt/src/integration/java/com/radixdlt/consensus/MultiNodeConsensusTest.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.commons.math3.util.Pair;
+import org.junit.Before;
+import org.junit.Test;
+import org.radix.logging.Logger;
+import org.radix.logging.Logging;
+
+public class MultiNodeConsensusTest {
+	private static final Logger log = Logging.getLogger("Test");
+	private final TestEventCoordinatorNetwork testEventCoordinatorNetwork = new TestEventCoordinatorNetwork();
+
+	private Atom genesis;
+	private Vertex genesisVertex;
+	private QuorumCertificate genesisQC;
+
+	@Before
+	public void setup() {
+		this.genesis = mock(Atom.class);
+		when(this.genesis.toString()).thenReturn(Long.toHexString(0));
+		this.genesisVertex = Vertex.createGenesis(genesis);
+		this.genesisQC = new QuorumCertificate(
+			new VertexMetadata(View.genesis(), genesisVertex.getId(), null, null),
+			new ECDSASignatures()
+		);
+	}
+
+	private ChainedBFT createBFTInstance(
+		ECKeyPair key,
+		ProposerElection proposerElection,
+		QuorumRequirements quorumRequirements,
+		VertexStore vertexStore
+	) {
+		Mempool mempool = mock(Mempool.class);
+		AtomicLong atomId = new AtomicLong();
+		doAnswer(inv -> {
+			Atom atom = mock(Atom.class);
+			when(atom.toString()).thenReturn(Long.toHexString(atomId.incrementAndGet()));
+			return Collections.singletonList(atom);
+		}).when(mempool).getAtoms(anyInt(), anySet());
+		ProposalGenerator proposalGenerator = new ProposalGenerator(vertexStore, mempool);
+		SafetyRules safetyRules = new SafetyRules(key, vertexStore, SafetyState.initialState());
+		PacemakerImpl pacemaker = new PacemakerImpl(quorumRequirements, Executors.newSingleThreadScheduledExecutor());
+		PendingVotes pendingVotes = new PendingVotes(quorumRequirements);
+		EventCoordinator eventCoordinator = new EventCoordinator(
+			proposalGenerator,
+			mempool,
+			testEventCoordinatorNetwork.getNetworkSender(key.getUID()),
+			safetyRules,
+			pacemaker,
+			vertexStore,
+			pendingVotes,
+			proposerElection,
+			key
+		);
+
+		return new ChainedBFT(
+			eventCoordinator,
+			testEventCoordinatorNetwork.getNetworkRx(key.getUID()),
+			pacemaker
+		);
+	}
+
+	private List<TestObserver<Vertex>> runBFT(
+		List<ECKeyPair> nodes,
+		QuorumRequirements quorumRequirements,
+		ProposerElection proposerElection
+	) {
+		final List<Pair<TestObserver<Vertex>, Observable<Event>>> bftEvents = nodes.stream()
+			.map(e -> {
+				RadixEngine radixEngine = mock(RadixEngine.class);
+				when(radixEngine.staticCheck(any())).thenReturn(Optional.empty());
+				VertexStore vertexStore = new VertexStore(genesisVertex, genesisQC, radixEngine);
+				TestObserver<Vertex> testObserver = TestObserver.create();
+				vertexStore.lastCommittedVertex().subscribe(testObserver);
+				ChainedBFT chainedBFT = createBFTInstance(e, proposerElection, quorumRequirements, vertexStore);
+				return new Pair<>(testObserver, chainedBFT.processEvents());
+			})
+			.collect(Collectors.toList());
+
+		Observable.merge(bftEvents.stream().map(Pair::getSecond).collect(Collectors.toList()))
+			.subscribe();
```

Comment:
Could we subscribe directly inside the stream instead of passing `chainedBFT.processEvents()` to the "outside" through `Pairs`?

Reply:
good idea